tests/test_main.py

```diff
@@ -80,3 +83,82 @@ def install_signal_handlers(self):
     response = requests.get("http://127.0.0.1:8000")
     assert response.status_code == 204
     thread.join()
+
+
+def test_run_signal():
+    class App:
+        def __init__(self, scope):
+            if scope["type"] != "http":
+                raise Exception()
+
+        async def __call__(self, receive, send):
+            await send({"type": "http.response.start", "status": 204, "headers": []})
+            await send({"type": "http.response.body", "body": b"", "more_body": False})
+
+    config = Config(app=App, loop="asyncio", limit_max_requests=1, handled_signals=None)
+    server = Server(config=config)
+    thread = threading.Thread(target=server.run)
+    thread.start()
+    while not server.started:
+        time.sleep(0.01)
+    response = requests.get("http://127.0.0.1:8000")
+    assert response.status_code == 204
+    thread.join()
+
+
+def test_run_signal_multi_threads():
+    class App:
+        def __init__(self, scope):
+            if scope["type"] != "http":
+                raise Exception()
+
+        async def __call__(self, receive, send):
+            await send({"type": "http.response.start", "status": 204, "headers": []})
+            await send({"type": "http.response.body", "body": b"", "more_body": False})
+
+    def join_t(*threads: threading.Thread) -> typing.List[None]:
+        return [t.join() for t in threads]
+
+    def start_threads(*threads: threading.Thread) -> typing.List[None]:
+        return [t.start() for t in threads]
+
+    def event_thread(
+        worker: typing.Awaitable, loop, *args, **kwargs
+    ) -> threading.Thread:
+        def _worker(*args, **kwargs):
+            try:
+                loop.run_until_complete(worker(*args, **kwargs))
+            except Exception as e:
+                print(e)
+            finally:
+                loop.close()
+
+        return threading.Thread(target=_worker, args=args, kwargs=kwargs)
+
+    threads_count = 10
+    loops = [asyncio.new_event_loop() for i in range(threads_count)]
+    for loop in loops:
+        asyncio.set_event_loop(loop)
+    configs = [
+        Config(
+            app=App,
+            port=10000 + i,
+            loop=loops[i],
+            limit_max_requests=1,
+            handled_signals=None,
+        )
+        for i in range(threads_count)
+    ]
+    servers = [Server(config=configs[i]) for i in range(threads_count)]
+    workers = [
+        event_thread(servers[i].serve, loop=loops[i]) for i in range(threads_count)
+    ]
+    start_threads(*workers)
+    time.sleep(1)
+    for x in range(threads_count):
+        port = 10000 + x
+        s = requests.Session()
+        s.mount(f"http://127.0.0.1:{port}", HTTPAdapter(max_retries=1))
+        response = s.get(f"http://127.0.0.1:{port}")
+        assert response.status_code == 204
+    join_t(*workers)
```

Comment:
I actually don't think we need to add these tests. We already have coverage of the code that's being used - that's enough for me here.

Reply:
fair enough

however I wanted to say that the main use case for this PR, if I understanded it correctly what people wanted, I mean using `handled_signals=None` is for people who may want a finer control over that Config part, so that they could run uvicorn in one thread and something else in another.

that test is effectively far-fetched in the sense it runs 10 servers in 10 threads, but it just demonstrates how it could be done,

skipping it removes the need for modifying the `tests/client.py` above since one-side effect of the test is that it as to close the loop!